src/WinRT.Runtime/ComWrappersSupport.cs

```diff
@@ -68,34 +68,44 @@ public static void MarshalDelegateInvoke<T>(IntPtr thisPtr, Action<T> invoke)
 
         // If we are free threaded, we do not need to keep track of context.
         // This can either be if the object implements IAgileObject or the free threaded marshaler.
-        internal unsafe static bool IsFreeThreaded(IObjectReference objRef)
+        internal unsafe static bool IsFreeThreaded(IntPtr iUnknown)
         {
-            if (objRef.TryAs(InterfaceIIDs.IAgileObject_IID, out var agilePtr) >= 0)
+            Guid iid_IAgileObject = InterfaceIIDs.IAgileObject_IID;
+            if (Marshal.QueryInterface(iUnknown, ref iid_IAgileObject, out var agilePtr) >= 0)
             {
                 Marshal.Release(agilePtr);
                 return true;
             }
-            else if (objRef.TryAs(InterfaceIIDs.IMarshal_IID, out var marshalPtr) >= 0)
+            else
             {
-                try
+                Guid iid_IMarshal = InterfaceIIDs.IMarshal_IID;
+                if (Marshal.QueryInterface(iUnknown, ref iid_IMarshal, out var marshalPtr) >= 0)
                 {
-                    Guid iid_IUnknown = InterfaceIIDs.IUnknown_IID;
-                    Guid iid_unmarshalClass;
-                    Marshal.ThrowExceptionForHR((**(ABI.WinRT.Interop.IMarshal.Vftbl**)marshalPtr).GetUnmarshalClass_0(
-                        marshalPtr, &iid_IUnknown, IntPtr.Zero, MSHCTX.InProc, IntPtr.Zero, MSHLFLAGS.Normal, &iid_unmarshalClass));
-                    if (iid_unmarshalClass == ABI.WinRT.Interop.IMarshal.IID_InProcFreeThreadedMarshaler.Value)
+                    try
                     {
-                        return true;
+                        Guid iid_IUnknown = InterfaceIIDs.IUnknown_IID;
+                        Guid iid_unmarshalClass;
+                        Marshal.ThrowExceptionForHR((**(ABI.WinRT.Interop.IMarshal.Vftbl**)marshalPtr).GetUnmarshalClass_0(
+                            marshalPtr, &iid_IUnknown, IntPtr.Zero, MSHCTX.InProc, IntPtr.Zero, MSHLFLAGS.Normal, &iid_unmarshalClass));
+                        if (iid_unmarshalClass == ABI.WinRT.Interop.IMarshal.IID_InProcFreeThreadedMarshaler.Value)
+                        {
+                            return true;
+                        }
+                    }
+                    finally
+                    {
+                        Marshal.Release(marshalPtr);
                     }
-                }
-                finally
-                {
-                    Marshal.Release(marshalPtr);
                 }
             }
             return false;
         }
 
+        internal unsafe static bool IsFreeThreaded(IObjectReference objRef)
+        {
+            return IsFreeThreaded(objRef.ThisPtr);

```

Comment:
Question on this, but really I noticed the same thing throughout CsWinRT. How is this safe without a `GC.KeepAlive` call? The GC could collect and finalize `objRef` while `IsFreeThreaded` is still running, which will cause things to explode (as the object would be deleted while some method is running). Does this just rely on callers keeping the references alive in some other way? Is this provably safe or is this just a "it's more of a theoretical problem but in practice we didn't see any crashes yet"?

Reply:
This would rely on the caller keeping the reference alive.

Reply:
In this case, as mentioned, the caller would probably keep it alive.  For the general problem, we do have a bug to go look at if we should add `GC.KeepAlive` in certain places but so far, we haven't seen any crashes arise yet from not having it.

Reply:
Added a GC.KeepAlive here.